### PR TITLE
Make some functions more strict

### PR DESCRIPTION
**Pull request by konsumlamm**
Make some functions more strict, simplify `Data.Mod.Word.^%` and fix a typo in `invertModWord`.

### CODE REVIEW

File: Data/Mod/Word.hs [hunk 1]
```diff
@@ -247,14 +250,14 @@ instance KnownNat m => Fractional (Mod m) where
 -- >>> invertMod 4 :: Mod 10 -- 4 and 10 are not coprime
 -- Nothing
 invertMod :: KnownNat m => Mod m -> Maybe (Mod m)
-invertMod mx@(Mod x) = case natVal mx of
+invertMod mx@(Mod !x) = case natVal mx of
```
Comment (Bodigrim) [inline Data/Mod/Word.hs:277]: `Mod` is a newtype, what's the reason to add a bang here?

File: Data/Mod/Word.hs [hunk 2]
```diff
@@ -345,22 +348,16 @@ half x = x `shiftR` 1
...
-      f :: Integral a => Word# -> a -> Word# -> Word#
-      f _  0 acc# = acc#
-      f b# e acc# = f bb# (e `P.quot` 2) (if odd e then ba# else acc#)
-        where
-          !(# bb1#, bb2# #) = timesWord2# b# b#
-          !(#    _, bb#  #) = quotRemWord2# bb1# bb2# m#
-          !(# ba1#, ba2# #) = timesWord2# b# acc#
-          !(#    _, ba#  #) = quotRemWord2# ba1# ba2# m#
+      f !_ 0 acc = acc
+      f b  e acc = f (mulMod m b b) (e `P.quot` 2) (if odd e then mulMod m b acc else acc)
```
Comment (Bodigrim) [inline Data/Mod/Word.hs:384]: Are you sure that there is no performance regression?

Comment (konsumlamm) [reply Data/Mod/Word.hs:384]: Pretty sure, the `Word`s still get unboxed and the generated core is even smaller (there are no dictionaries being created every iteration anymore). Additionally, `mulMod m b acc` is now only computed when it is needed.

File: Data/Mod/Word.hs (hunk 1 above)
Comment (konsumlamm) [reply Data/Mod/Word.hs:277]: It's a newtype, so matching the constructor doesn't force the value. Adding the bang causes GHC to generate a version which takes a `Word#`, which gets used for calls to `invertMod` which previously passed a boxed `Word`.

**Comment by Bodigrim** [comment]
(I'm not convinced about changes to `quotRem`, but I'm going to rewrite them in course of #20 anyway)

**Pull request merged**